// twitter-screen-scrape/lib/index.js
// Generated by CoffeeScript 1.10.0
(function() {
  var ACTIONS, Readable, TwitterPosts, checkStatus, cheerio, fetch, getPostElements, queryString,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Readable = require('readable-stream/readable');

  cheerio = require('cheerio');

  fetch = require('node-fetch');

  queryString = require('querystring');

  ACTIONS = ['reply', 'retweet', 'favorite'];

  checkStatus = function(response) {
    var error;
    if (response.status >= 200 && response.status < 300) {
      return response;
    } else {
      error = new Error(response.statusText);
      error.response = response;
      throw error;
    }
  };


  /**
   * Make a request for a Twitter page, parse the response, and get all the tweet
     elements.
   * @param {String} username
   * @param {String} [startingId] The maximum tweet id query for (the lowest one
     from the last request), or undefined if this is the first request.
   * @return {Array} An array of elements.
   */

  getPostElements = function(username, startingId) {
    var options, url;
    url = "https://twitter.com/i/profiles/show/" + username + "/timeline";
    options = {
      'include_available_features': '1',
      'include_entities': '1'
    };
    if (startingId != null) {
      options['max_position'] = startingId;
    }
    url += '?' + queryString.stringify(options);
    return fetch(url).then(checkStatus).then(function(response) {
      return response.json();
    });
  };


  /**
   * Stream that scrapes as many tweets as possible for a given user.
   * @param {String} options.username
   * @param {Boolean} options.retweets Whether to include retweets.
   * @return {Stream} A stream of tweet objects.
   */

  TwitterPosts = (function(superClass) {
    extend(TwitterPosts, superClass);

    TwitterPosts.prototype._lock = false;

    TwitterPosts.prototype._minPostId = void 0;

    TwitterPosts.prototype._lastMinPostId = void 0;

    function TwitterPosts(arg) {
      this.username = arg.username, this.retweets = arg.retweets;
      this.destroy = bind(this.destroy, this);
      this._read = bind(this._read, this);
      if (this.retweets == null) {
        this.retweets = true;
      }
      TwitterPosts.__super__.constructor.call(this, {
        highWaterMark: 16,
        objectMode: true
      });
      this._readableState.destroyed = false;
    }

    TwitterPosts.prototype._read = function() {
      var lastPost;
      if (this._lock) {
        return;
      }
      this._lock = true;
      if (this._readableState.destroyed) {
        this.push(null);
        return;
      }
      lastPost = void 0;
      return getPostElements(this.username, this._minPostId).then(function(response) {
        var html;
        html = response['items_html'].trim();
        return cheerio.load(html);
      }).then((function(_this) {
        return function($) {
          var action, element, hasEmitted, hasMorePosts, id, isRetweet, j, k, l, len, len1, len2, pic, pics, post, ref, textElement, wrapper;
          hasEmitted = false;
          ref = $('.original-tweet');
          for (j = 0, len = ref.length; j < len; j++) {
            element = ref[j];
            id = $(element).attr('data-item-id');
            _this._minPostId = id;
            isRetweet = $(element).find('.js-retweet-text').length !== 0;
            if (!_this.retweets && isRetweet) {
              continue;
            }
            textElement = $(element).find('.tweet-text').first();
            textElement.find('img.twitter-emoji').each(function(i, emoji) {
              return $(emoji).html($(emoji).attr('alt'));
            });
            post = {
              id: id,
              isRetweet: isRetweet,
              username: _this.username,
              text: textElement.text(),
              time: +$(element).find('.js-short-timestamp').first().attr('data-time'),
              images: []
            };
            for (k = 0, len1 = ACTIONS.length; k < len1; k++) {
              action = ACTIONS[k];
              wrapper = $(element).find(".ProfileTweet-action--" + action + " .ProfileTweet-actionCount");
              post[action] = (wrapper.length !== 0 ? +$(wrapper).first().attr('data-tweet-stat-count') : void 0);
            }
            pics = $(element).find('.multi-photos .multi-photo[data-image-url], [data-card-type=photo] [data-image-url]');
            for (l = 0, len2 = pics.length; l < len2; l++) {
              pic = pics[l];
              post.images.push($(pic).attr('data-image-url'));
            }
            if (lastPost != null) {
              _this.push(lastPost);
              hasEmitted = true;
            }
            lastPost = post;
          }
          hasMorePosts = _this._lastMinPostId !== _this._minPostId;
          if (hasMorePosts) {
            _this._lock = false;
            _this._lastMinPostId = _this._minPostId;
          }
          if (lastPost != null) {
            _this.push(lastPost);
            hasEmitted = true;
          }
          if (!hasMorePosts) {
            _this.push(null);
          }
          if (!hasEmitted && hasMorePosts) {
            return _this._read();
          }
        };
      })(this))["catch"]((function(_this) {
        return function(err) {
          return _this.emit('error', err);
        };
      })(this));
    };

    TwitterPosts.prototype.destroy = function() {
      if (this._readableState.destroyed) {
        return;
      }
      this._readableState.destroyed = true;
      return this._destroy((function(_this) {
        return function(err) {
          if (err) {
            _this.emit('error', err);
          }
          return _this.emit('close');
        };
      })(this));
    };

    TwitterPosts.prototype._destroy = function(cb) {
      return process.nextTick(cb);
    };

    return TwitterPosts;

  })(Readable);

  module.exports = TwitterPosts;

}).call(this);
